src/documentation/programming-guide.md

```diff
@@ -1372,119 +1923,228 @@ public class MyCustomDataType {
+A **session window** function defines windows that contain elements that are
+within a certain gap duration of another element. Session windowing applies on a
+per-key basis and is useful for data that is irregularly distributed with
+respect to time. For example, a data stream representing user mouse activity may
+have long periods of idle time interspersed with high concentrations of clicks.
+If data arrives after the minimum specified gap duration time, this initiates
+the start of a new window.
 
 ![Diagram of session windows with a minimum gap duration]({{ "/images/session-windows.png" | prepend: site.baseurl }} "Session windows, with a minimum gap duration")
 
-**Figure:** Session windows, with a minimum gap duration. Note how each data key has different windows, according to its data distribution.
+**Figure:** Session windows, with a minimum gap duration. Note how each data key
+has different windows, according to its data distribution.
 
-#### Single global window
+#### 7.2.4. Single global window
 
-By default, all data in a `PCollection` is assigned to a single global window, and late data is discarded. If your data set is of a fixed size, you can use the global window default for your `PCollection`.
+By default, all data in a `PCollection` is assigned to a single global window,
+and late data is discarded. If your data set is of a fixed size, you can use the
+global window default for your `PCollection`.
 
-You can use a single global window if you are working with an unbounded data set (e.g. from a streaming data source) but use caution when applying aggregating transforms such as `GroupByKey` and `Combine`. A single global window with a default trigger generally requires the entire data set to be available before processing, which is not possible with continuously updating data. To perform aggregations on an unbounded `PCollection` that uses global windowing, you should specify a non-default trigger for that `PCollection`.
+You can use a single global window if you are working with an unbounded data set
+(e.g. from a streaming data source) but use caution when applying aggregating
+transforms such as `GroupByKey` and `Combine`. A single global window with a
+default trigger generally requires the entire data set to be available before
+processing, which is not possible with continuously updating data. To perform
+aggregations on an unbounded `PCollection` that uses global windowing, you
+should specify a non-default trigger for that `PCollection`.
 
-### <a name="setwindowingfunction"></a>Setting your PCollection's windowing function
+### 7.3. Setting your PCollection's windowing function
```

Comment:
Can we condense the "How to construct a window fn" to the above section, and have this be just how to apply a WindowFn to a PCollection?

(Feel free to JIRA and do later)

Reply:
opened JIRA